ocs_ci/ocs/cluster.py

```diff
@@ -427,6 +431,86 @@ def check_ceph_pool_used_space(self, cbp_name):
             f"In Rados df, Used size is varying"
         )
 
+    def get_ceph_health(self, detail=False):
+        """
+        Exec `ceph health` cmd on tools pod and return the status of the ceph
+        cluster.
+
+        Args:
+            detail (bool): If True the 'ceph health detail' is executed
+
+        Returns:
+            str: Output of the ceph health command.
+
+        """
+        ceph_health_cmd = "ceph health"
+        if detail:
+            ceph_health_cmd = f"{ceph_health_cmd} detail"
+
+        return self.toolbox.exec_cmd_on_pod(
+            ceph_health_cmd, out_yaml_format=False,
+        )
+
+    def get_ceph_status(self):
+        """
+        Exec `ceph status` cmd on tools pod and return its output.
+
+        Returns:
+            str: Output of the ceph status command.
+
+        """
+        return self.toolbox.exec_cmd_on_pod(
+            "ceph status", out_yaml_format=False,
+        )
+
+    def enable_health_monitor(self, sleep=5):
+        """
+        Enable monitoring for ceph health status.
+
+        Args:
+            sleep (int): Number of seconds to sleep between health checks.
+
+        """
+        self.monitor = HealthMonitorThread(self, sleep)
+        self.monitor.start()
+
+    def disable_health_monitor(self):
+        self.health_monitor_enabled = False
+
+
+class HealthMonitorThread(threading.Thread):
+    """
+    Class for monitoring ceph health status of CephCluster. If CephCluster will
+    get to HEALTH_ERROR state it will save the ceph status to
+    health_error_status variable in ceph_cluster object and will stop
+    monitoring. It's up to user how to handle the error.
+    """
+
+    def __init__(self, ceph_cluster, sleep=5):
+        """
+        Constructor for ceph health status thread.
+
+        Args:
+            ceph_cluster (CephCluster): Reference to CephCluster object.
+            sleep (int): Number of seconds to sleep between health checks.
+
+        """
+        self.ceph_cluster = ceph_cluster
+        self.sleep = sleep
+        super(HealthMonitorThread, self).__init__()
+
+    def run(self):
+        self.ceph_cluster.health_monitor_enabled = True
+        while self.ceph_cluster.health_monitor_enabled and (
+            not self.ceph_cluster.health_error_status
+        ):
+            sleep(self.sleep)
+            health_status = self.ceph_cluster.get_ceph_health(detail=True)
+            if "HEALTH_ERROR" in health_status:
+                self.ceph_cluster.health_error_status = (
```

Comment:
This means that only 1 error occurrence will be assigned to self.ceph_cluster.health_error_status IIUC.
What if we want more then 1 error occurrence to be saved?

Reply:
I was thinking about adding more as well but then I realized that if I run this health check each 5 second and I will hit HEALTH_ERROR and this one will repeat all the time that I will spam this list with the same error all the time and no more additional error.  I thought about concentrating for first occurrence of HEALTH_ERROR only and stop monitoring after that, let process the rest of the upgrade which will most likely fail later on anyway and collect must gather and raise the exception with that first error.

@nehaberry or do you think that we should try to collect all the errors again and again?

Thanks

Reply:
@petr-balogh OK. Make sense to have it saved only once. Thanks!